what-is-chalk.mdx

```diff
@@ -18,215 +18,440 @@ import { PyDiffEditor } from '@/components/Editor'
+
+## Chalk for Data Scientists
+
+Test new features, run experiments, and ship production models—all from a Jupyter notebook.
+
+- catch regressions and A/B test features on development branches
+- import features from production into Jupyter notebooks with a single line of code
+- export features to catalogs like Iceberg for downstream analytics and usage
+- trace data lineage all the way down to source tables
+
+```python
+from chalk.client import ChalkClient
+
+client = ChalkClient(branch="elvis")
+client.load_features() # load prod features with one line
+
+User.name_exclaimed = _.name + "!" # add new features
+
+chalk_dataset = client.offline_query(
+    input={
+        User.id: list(range(1000)),
+    },
+    output={
+        User.id,
+        User.name,
+        User.name_exclaimed,
+    },
+    recompute_features=True, # A/B test against historical model runs
+    run_asynchronously=True,
+    dataset_name="<dataset_name>",
+)
+
+df = chalk_dataset.to_pandas() # convert to pandas dataframe
+
+# write to Glue or Iceberg
+catalog = GlueCatalog(
+    name="aws_glue_catalog",
+    aws_region="us-west-2",
+    catalog_id="123",
+    aws_role_arn="arn:aws:iam::123456789012:role/OurCatalogueAccessRole",
+)
+chalk_dataset.write_to(destination="database.table_name", catalog=catalog)
+```
+
+Chalk evaluates features with point-in-time lookups guaranteeing evaluation only with data that would have been seen in the past, this means that you can provide labels with different past timestamps to easily get historical features that represent what your application would have retrieved online at those past times.
 
 <TemporalCorrectness />
+
+With Chalk, Data scientists easily integrate new data sources, prototype with guaranteed point-in-time correctness, and collaborate with branches and data exports.
+
+## Chalk for Data Engineers
+
+Manage features declaratively--with a few lines of code--all without the complexity of orchestrating ETL jobs, feature stores, and offline/online datastores.
+
+- cache expensive computations with configurable staleness
+- manage versioning and A/B testing of feature computation
+- create complex joins and relationships between data entities
+- configure time-window aggregations with flexible materialization options
+
+```python
+@features
+class User:
+    id: int
+    domain: str
+
+    # composite keys that can be used as join keys
+    workspace_id: str = _.domain + "-" + _.id
+    expensive_api_call: str = feature(max_staleness="30d") # cache values
+
+    # maintain different resolvers to A/B test function calls e.g. gemini vs openai
+    llm_response: str = feature(version=3)
+
+    # multi-attribute joins
+    txns: DataFrame[Transaction] = has_many(
+        lamda: (User.id == Transaction.user_id) & (User.domain == Transaction.domain)
+    )
```

Comment:
this hardly works. I tihnk we should stick to a single key join -- seems fine to have only transactions joined on user.

Reply:
switched to signle key